File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/MissingRuntimeMemberDiagnosticHelper.cs
using System.Collections.Generic;
using Microsoft.CodeAnalysis.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace Microsoft.CodeAnalysis
{
	[StandardModule]
	internal sealed class MissingRuntimeMemberDiagnosticHelper
	{
		internal const string MyVBNamespace = "My";

		private static readonly Dictionary<string, string> s_metadataNames = new Dictionary<string, string>
		{
			{ "Microsoft.VisualBasic.CompilerServices.Operators", "Late binding" },
			{ "Microsoft.VisualBasic.CompilerServices.NewLateBinding", "Late binding" },
			{ "Microsoft.VisualBasic.CompilerServices.LikeOperator", "Like operator" },
			{ "Microsoft.VisualBasic.CompilerServices.ProjectData", "Unstructured exception handling" },
			{ "Microsoft.VisualBasic.CompilerServices.ProjectData.CreateProjectError", "Unstructured exception handling" }
		};

		internal static DiagnosticInfo GetDiagnosticForMissingRuntimeHelper(string typename, string membername, bool embedVBCoreRuntime)
		{
			string value = "";
			s_metadataNames.TryGetValue(typename, out value);
			if (embedVBCoreRuntime && !string.IsNullOrEmpty(value))
			{
				return ErrorFactory.ErrorInfo(ERRID.ERR_PlatformDoesntSupport, value);
			}
			return ErrorFactory.ErrorInfo(ERRID.ERR_MissingRuntimeHelper, typename + "." + membername);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/MissingRuntimeMemberDiagnosticHelper.vb
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.VisualBasic.CompilerServices
Imports System
Imports System.Collections.Generic

Namespace Microsoft.CodeAnalysis
	Friend Module MissingRuntimeMemberDiagnosticHelper
		Friend Const MyVBNamespace As String = "My"

		Private ReadOnly s_metadataNames As Dictionary(Of String, String)

		Sub New()
			MissingRuntimeMemberDiagnosticHelper.s_metadataNames = New Dictionary(Of String, String)() From
			{
				{ "Microsoft.VisualBasic.CompilerServices.Operators", "Late binding" },
				{ "Microsoft.VisualBasic.CompilerServices.NewLateBinding", "Late binding" },
				{ "Microsoft.VisualBasic.CompilerServices.LikeOperator", "Like operator" },
				{ "Microsoft.VisualBasic.CompilerServices.ProjectData", "Unstructured exception handling" },
				{ "Microsoft.VisualBasic.CompilerServices.ProjectData.CreateProjectError", "Unstructured exception handling" }
			}
		End Sub

		Friend Function GetDiagnosticForMissingRuntimeHelper(ByVal typename As String, ByVal membername As String, ByVal embedVBCoreRuntime As Boolean) As Microsoft.CodeAnalysis.DiagnosticInfo
			Dim diagnosticInfo As Microsoft.CodeAnalysis.DiagnosticInfo
			Dim str As String = ""
			MissingRuntimeMemberDiagnosticHelper.s_metadataNames.TryGetValue(typename, str)
			diagnosticInfo = If(Not embedVBCoreRuntime OrElse [String].IsNullOrEmpty(str), ErrorFactory.ErrorInfo(ERRID.ERR_MissingRuntimeHelper, New [Object]() { [String].Concat(typename, ".", membername) }), ErrorFactory.ErrorInfo(ERRID.ERR_PlatformDoesntSupport, New [Object]() { str }))
			Return diagnosticInfo
		End Function
	End Module
End Namespace